src/bfmi.jl

```diff
@@ -0,0 +1,41 @@
+"""
+    bfmi(energy; dims=1)
+
+Calculate the estimated Bayesian fraction of missing information (BFMI).
+
+When sampling with Hamiltonian Monte Carlo (HMC), BFMI quantifies how well momentum
+resampling matches the marginal energy distribution.
+
+The current advice is that values smaller than 0.3 indicate poor sampling.
+However, this threshold is provisional and may change.
+A BFMI value below the threshold often indicates poor adaptation of sampling parameters or
+that the target distribution has heavy tails that were not well explored by the Markov
+chain.
+
+For more information, see Section 6.1 of [^Betancourt2018] or [^Betancourt2016] for a
+complete account.
+
+`energy` is either a vector of Hamiltonian energies of draws or a matrix of energies of
+draws for multiple chains.
+`dims` indicates the dimension in `energy` that contains the draws.
+The default `dims=1` assumes `energy` has the shape `draws` or `(draws, chains)`.
+If a different shape is provided, `dims` must be set accordingly.
+
+If `energy` is a vector, a single BFMI value is returned.
+Otherwise, a vector of BFMI values for each chain is returned.
+
+# References
+
+[^Betancourt2018]: Betancourt M. (2018).
+    A Conceptual Introduction to Hamiltonian Monte Carlo.
+    [arXiv:1701.02434v2](https://arxiv.org/pdf/1701.02434v2.pdf) [stat.ME]
+[^Betancourt2016]: Betancourt M. (2016).
+    Diagnosing Suboptimal Cotangent Disintegrations in Hamiltonian Monte Carlo.
+    [arXiv:1604.00695v1](https://arxiv.org/pdf/1604.00695v1.pdf) [stat.ME]
+"""
+function bfmi(energy; dims=1)
```

Comment:
Should this be restricted to something like
```suggestion
function bfmi(energy::AbstractArray{<:Real}; dims=1)
```
as other methods?

Reply:
And do we want to support tuples, arrays, etc. for `dims` or rather restrict it to `dims::Int`?

Reply:
> Should this be restricted to something like
> 
> as other methods?

For consistency, probably.



> And do we want to support tuples, arrays, etc. for `dims` or rather restrict it to `dims::Int`?

I don't see a use case for tuple or array dims (inputs must either be a vector of draws or a matrix of draws for multiple chains). However, I don't think we should restrict to `Int` dims. e.g. for dimensional arrays, we can pass a specialized dimension object to select a draw dimension by name. e.g. from the ArviZ example:

```julia
julia> bfmi(energy; dims=:draw)
4-element DimArray{Float64,1} with dimensions: 
  Dim{:chain} Sampled 0:3 ForwardOrdered Regular Points
and reference dimensions: 
  Dim{:draw} Sampled 249.5:500.0:249.5 ForwardOrdered Regular Points
 0.289451
 0.409332
 0.4406
 0.235864

julia> bfmi(permutedims(energy); dims=:draw)
4-element DimArray{Float64,1} with dimensions: 
  Dim{:chain} Sampled 0:3 ForwardOrdered Regular Points
and reference dimensions: 
  Dim{:draw} Sampled 249.5:500.0:249.5 ForwardOrdered Regular Points
 0.289451
 0.409332
 0.4406
 0.235864
```

Reply:
> However, I don't think we should restrict to `Int` dims. e.g. for dimensional arrays, we can pass a specialized dimension object to select a draw dimension by name. e.g. from the ArviZ example:

Wouldn't we always want to use `:draw` in that example? I.e., wouldn't it be better to add a dispatch for `DimArray` that always picks `:draw` (at least if the dimensions contain `Dim{:draw}` which - I assume, not familiar with that package - could be determined at compile time) instead of supporting arbitrary `dims`?

Reply:
If `DimArray` was owned by ArviZ, yes, but since it isn't, that would be type piracy. e.g. someone else could use DimensionalData to store MCMC draws independently of ArviZ and name the draw dimension `:samples`. In practice, we would define a method like:

```julia
function MCMCDiagnosticTools.bfmi(idata::InferenceData)
    energy = idata.sample_stats.energy
    dims = Dimensions.dims(energy, :draw)
    MCMCDiagnosticTools.bfmi(energy; dims)
end
```
which we can do in ArviZ because `:draw` is a reserved name for MCMC draw dimensions. Another perfectly valid option uses `Dimensions.dimnum(energy, :draw)` to get the integer dimension. Both resolve at compile time (I assume due to inlining).

What problems do you see with supporting arbitrary `dims`? We're just forwarding them to other methods, so if other methods support the provided dims, then so would we. This seems to me to be preferable. By not setting type constraints, we allow for more general and composable code. e.g. `LogExpFunctions.logsumexp` just works for DimensionalData since it puts no constraints on `dims` and instead just forwards them to `reduce`.

Reply:
My main concern here is that in my experience unrestricted `dims` keyword arguments usually/often also support tuples, arrays, etc. of dimensions but we don't want that here (in contrast to e.g. `logsumexp`, and therefore I think the comparison is a bit flawed). The simplest way to avoid these issues would be to just restrict `dims` to `dims::Int` for now. That seems to work for the usual inputs in MCMCDiagnosticTools (regular arrays) and would also work for structures such as `InferenceData` or `MCMCChains.Chain` by forwarding calls to the underlying wrapped array (for which `dims::Int` would be sufficient)?

Reply:
While I disagree that this is a problem here (we don't use `dims`, we just forward it and let other methods error if what is requested is unreasonable), it's true that in ArviZ and MCMCChains (I assume) we can make do with an integer constraint, and I'm okay with this compromise.

Reply:
Somehow I tend towards the more conservative approach here and think `dims::Int` would be a good start. We can easily remove the type annotation if it turns out that it is too restrictive.

Reply:
Done!